neuromation/api/images.py

```diff
@@ -140,20 +144,65 @@ def _auth(self) -> Dict[str, str]:
             raise  # pragma: no cover
 
         async for obj in stream:
-            if "error" in obj.keys():
-                error_details = obj.get("errorDetail", {"message": "Unknown error"})
-                raise DockerError(900, error_details)
-            elif "id" in obj.keys() and obj["id"] != remote.tag:
-                if "progress" in obj.keys():
-                    message = f"{obj['id']}: {obj['status']} {obj['progress']}"
-                else:
-                    message = f"{obj['id']}: {obj['status']}"
-                progress.step(ImageProgressStep(message, obj["id"]))
+            step = _try_parse_image_progress_step(obj, remote.tag)
+            if step:
+                progress.step(step)
 
         await self._docker.images.tag(repo, local)
 
         return local
 
+    async def save(
+        self,
+        id: str,
+        image: RemoteImage,
+        *,
+        container_progress: Optional[AbstractDockerContainerProgress] = None,
+        image_progress: Optional[AbstractDockerImageProgress] = None,
+    ) -> None:
+        if not _is_in_neuro_registry(image):
+            raise ValueError(f"Image `{image}` must be in the neuromation registry")
+        if container_progress is None:
+            container_progress = _DummyContainerProgress()
+        if image_progress is None:
+            image_progress = _DummyImageProgress()
+
+        payload = {"container": {"image": _as_repo_str(image)}}
+        url = self._config.cluster_config.monitoring_url / f"{id}/save"
+
+        async with self._core.request("POST", url, json=payload) as resp:
+
+            # first, we process docker-commit messages
+            container_progress.commit(ContainerProgressCommit(id, image))
+            async for chunk in resp.content:
```

Comment:
IMO cycle over docker-commit messages looks better than manual processing of 2 commit chunks:
```
# first, we expect exactly two docker-commit messages
container_progress.commit(ContainerProgressCommit(id, image))

commit_chunk_1 = self._load_chunk(await resp.content.readline())
step_1 = _try_parse_job_commit_progress_step(commit_chunk_1)
container_progress.step(step_1)

commit_chunk_2 = self._load_chunk(await resp.content.readline())
step_2 = _try_parse_job_commit_progress_step(commit_chunk_2)
container_progress.step(step_2)
if step_2.message != "Committed":
    error_details = {
        "message": f"Failed to commit: last chunk status: `{step_2.message}`"
    }
    raise DockerError(900, error_details)
...
```